kernel/src/log_segment.rs

```diff
@@ -154,4 +390,83 @@ mod tests {
         // missing columns, but can still skip part 3 because has valid nullcount stats for P&M.
         assert_eq!(data.len(), 4);
     }
+
+    #[test]
+    fn test_read_log_with_out_of_date_last_checkpoint() {
+        let store = Arc::new(InMemory::new());
+
+        fn get_path(index: usize, suffix: &str) -> Path {
+            let path = format!("_delta_log/{index:020}.{suffix}");
+            Path::from(path.as_str())
+        }
+        let data = bytes::Bytes::from("kernel-data");
+
+        let checkpoint_metadata = CheckpointMetadata {
+            version: 3,
+            size: 10,
+            parts: None,
+            size_in_bytes: None,
+            num_of_add_files: None,
+            checkpoint_schema: None,
+            checksum: None,
+        };
+
+        // add log files to store
+        tokio::runtime::Runtime::new()
+            .expect("create tokio runtime")
+            .block_on(async {
+                for path in [
+                    get_path(0, "json"),
+                    get_path(1, "checkpoint.parquet"),
+                    get_path(2, "json"),
+                    get_path(3, "checkpoint.parquet"),
+                    get_path(4, "json"),
+                    get_path(5, "checkpoint.parquet"),
+                    get_path(6, "json"),
+                    get_path(7, "json"),
+                ] {
+                    store
+                        .put(&path, data.clone().into())
+                        .await
+                        .expect("put log file in store");
+                }
+                let checkpoint_str =
+                    serde_json::to_string(&checkpoint_metadata).expect("Serialize checkpoint");
+                store
+                    .put(
+                        &Path::from("_delta_log/_last_checkpoint"),
+                        checkpoint_str.into(),
+                    )
+                    .await
+                    .expect("Write _last_checkpoint");
+            });
+
+        let client = ObjectStoreFileSystemClient::new(
+            store,
+            false, // don't have ordered listing
+            Path::from("/"),
+            Arc::new(TokioBackgroundExecutor::new()),
+        );
+
+        let url = Url::parse("memory:///_delta_log/").expect("valid url");
+        let (mut commit_files, checkpoint_files) =
+            LogSegmentBuilder::list_log_files_with_checkpoint(&checkpoint_metadata, &client, &url)
+                .unwrap();
+
+        // Make the most recent commit the first in iterator
+        commit_files.reverse();
```

Comment:
I wanted to flag this section. We are effectively doing the same work as the `build` function would do. However, `build` maps the `ParsedLogPath` to `file.location`. We need more information in this test than location, so we can't `build` and make the same assertions. 

I considered factoring out part of `build` into a `process_files` function, but it didn't seem worth adding an entire extra function when this is the only test that needs it.

Reply:
and i suppose if we have `LogSegment` hold `ParsedLogPath`s then this is moot?

Reply:
Correct. Just updated the code to use `ParsedLogPath` and update the test.